### DISCUSSION TITLE
TypeScript error when using RTK Matcher function with listenerMiddleware

### DISCUSSION THREAD

**Discussion by moloko** [reduxjs/redux-toolkit#4582] [Q&A] [answered]
So, I've got my listener middleware set up like this:
![image](https://github.com/user-attachments/assets/36ad05cb-cf46-48c1-8603-5d67bf261f24)

But although the code is working fine I'm getting a TypeScript error 'No overload matches this call' for both `matcher` and `effect`.

Apologies as I am quite new to both react-toolkit and TypeScript so I am not sure if I'm doing something wrong or there's a problem with the type definitions. I did upgrade to react-toolkit v2.2.7 but that doesn't seem to have helped...

**Comment by markerikson** [comment, marked as answer]
We've found that the RTK types sometimes have issues when you do a function call on the right side of an object field, like `configureStore({reducer: createRootReducer()})`, or in this case, `matcher: isAnyOf()`. 

Try moving that to a separate variable, like `const isElementAction = isAnyOf(setElementSTarted, setElementCompleted)`, and see if it fixes the issue.

Also, while it's not directly related, note that we designed the listener middleware API with the assumption that you'd write the `effect` functions inline to get the proper TS type inference, rather than defining them separately - any particular reason you defined the `effect` in another file here?

You might want to consider using the "setup function in slice" organization approach shown here:

- https://redux-toolkit.js.org/api/createListenerMiddleware#organizing-listeners-in-files
- https://redux.js.org/tutorials/essentials/part-6-performance-normalization#showing-toasts-for-new-posts

**Reply by moloko** [reply]
> Try moving that to a separate variable, like `const isElementAction = isAnyOf(setElementSTarted, setElementCompleted)`, and see if it fixes the issue.

It did indeed - thank you!

> Also, while it's not directly related, note that we designed the listener middleware API with the assumption that you'd write the `effect` functions inline to get the proper TS type inference, rather than defining them separately - any particular reason you defined the `effect` in another file here?

Not really, we're still very much at the 'figuring things out' stage of the work we're doing - so the suggestion & links you have provided are very helpful, thank you!
